### PR TITLE
parity, ulong and description

### PR DESCRIPTION
**Pull request by scottf**
Keeping up with changes to java
Issue 32 stream and consumer config descriptions

### CODE REVIEW

File: src/NATS.Client/Internals/Validator.cs [hunk 1]
```diff
@@ -39,26 +40,43 @@ internal static string ValidateDurableRequired(string durable, ConsumerConfigura
...
 
         public static String ValidateJetStreamPrefix(String s) {
             return ValidatePrintableExceptWildGtDollar(s, "Prefix", false);
         }
 
-        public static string ValidatePrintable(string s, String label , bool required)
+        public static string ValidateMaxLength(String s, int maxLength, bool required, String label) {
+            return Validate(s, required, label, () =>
+            {
+                int len = Encoding.UTF8.GetBytes(s).Length;
```
Comment (sixlettervariables) [inline src/NATS.Client/Internals/Validator.cs]:
```suggestion
                int len = Encoding.UTF8.GetByteCount(s);
```

File: src/NATS.Client/Internals/Validator.cs [hunk 2]
```diff
@@ -39,26 +40,43 @@ internal static string ValidateDurableRequired(string durable, ConsumerConfigura
...
+                if (len > maxLength) {
+                    throw new ArgumentException($"{label} cannot be longer than {maxLength} bytes but was {len} bytes");
+                }
+                return s;
+            });
+        }
+
+        public static string ValidateDescription(String s, bool required)
+        {
+            return ValidateMaxLength(s, 4096, required, "Description");
```
Comment (sixlettervariables) [inline src/NATS.Client/Internals/Validator.cs]: Can we hoist `4096` into a named constant?

File: src/NATS.Client/JetStream/StreamConfiguration.cs [hunk 3]
```diff
@@ -168,308 +174,323 @@ public StreamConfigurationBuilder(StreamConfiguration sc)
...
+            /// </summary>
+            /// <param name="subjects">the stream's subjects</param>
+            /// <returns>The StreamConfigurationBuilder</returns>
+            public StreamConfigurationBuilder WithSubjects(List<string> subjects) {
+                _subjects.Clear();
+                return AddSubjects(subjects);
+            }
+
+            /// <summary>
+            /// Sets the subjects in the StreamConfiguration.
```
Comment (ColinSullivan1) [inline src/NATS.Client/JetStream/StreamConfiguration.cs]: Adds subjects to the StreamConfiguration

File: src/NATS.Client/JetStream/StreamConfiguration.cs [hunk 4]
```diff
@@ -168,308 +174,323 @@ public StreamConfigurationBuilder(StreamConfiguration sc)
...
-        /// </summary>
-        /// <param name="subjects">the stream's subjects</param>
-        /// <returns>The StreamConfigurationBuilder</returns>
-        public StreamConfigurationBuilder AddSubjects(List<string> subjects) {
-            if (subjects != null) {
-                foreach (string sub in subjects) {
-                    if (sub != null && !_subjects.Contains(sub)) {
-                        _subjects.Add(sub);
+            /// <summary>
+            /// Sets the subjects in the StreamConfiguration.
```
Comment (ColinSullivan1) [inline src/NATS.Client/JetStream/StreamConfiguration.cs]: See above...

File: src/NATS.Client/JetStream/StreamConfiguration.cs [hunk 5]
```diff
@@ -168,308 +174,323 @@ public StreamConfigurationBuilder(StreamConfiguration sc)
...
-            _noAck = noAck;
-            return this;
-        }
+            /// <summary>
+            /// Sets the acknowledgement mode of the StreamConfiguration.  if no acknowledgements are
+            /// set, then acknowledgements are not sent back to the client.  The default is false.
+            /// </summary>
+            /// <param name="noAck">true to disable acknowledgements.</param>
+            /// <returns>The StreamConfigurationBuilder</returns>
+            public StreamConfigurationBuilder WithNoAck(bool noAck) {
```
Comment (ColinSullivan1) [inline src/NATS.Client/JetStream/StreamConfiguration.cs:348]: It'd make more sense to do `WithAck(bool Ack)` and have the default be false.  Or is this mirroring go?

File: src/NATS.Client/Internals/Validator.cs (hunk 1 above)
Comment (scottf) [reply src/NATS.Client/Internals/Validator.cs]: fixed

File: src/NATS.Client/Internals/Validator.cs (hunk 2 above)
Comment (scottf) [reply src/NATS.Client/Internals/Validator.cs]: This function is being removed. I was told not to check the length on the client side

File: src/NATS.Client/JetStream/StreamConfiguration.cs (hunk 3 above)
Comment (scottf) [reply src/NATS.Client/JetStream/StreamConfiguration.cs]: fixed

File: src/NATS.Client/JetStream/StreamConfiguration.cs (hunk 4 above)
Comment (scottf) [reply src/NATS.Client/JetStream/StreamConfiguration.cs]: fixed

File: src/NATS.Client/JetStream/StreamConfiguration.cs (hunk 5 above)
Comment (scottf) [reply src/NATS.Client/JetStream/StreamConfiguration.cs:348]: I don't know what go is doing, I'll ask Wally. The feature is NoAck, and there are "negative" Options that work the same way. Also that's what I did in Java, not that that is right. I'll put a pin in this for now and come back to it.

**Pull request merged**